### DISCUSSION TITLE
Asynchronous "Action" operation?

### DISCUSSION THREAD

**Discussion by devEcommercePL** [filamentphp/filament#10627] [Help]
Is there any way to force "Action" to work asynchronously? Currently, if I click quickly on the action buttons, I get the result only after all actions are completed, not after each action.

```
                                Action::make('import')
                                    ->action(function (Set $set, $state) {
                                        dump(Carbon::now()->toJSON());
                                        sleep(3);
                                    })
```

**Comment by leandrocfe** [comment]
Why not jobs?

https://laravel.com/docs/10.x/queues#creating-jobs

**Reply by devEcommercePL** [reply]
Because I am using this to fetch data from an external API. After downloading, I need to update (real time) form and repeater with the downloaded data.

**Reply by devEcommercePL** [reply]
Any idea for another way to perform the action?

**Reply by leandrocfe** [reply]
Share some code you are trying to achieve

**Reply by devEcommercePL** [reply]
return Action::make("replace-image")
            ->action(function(Repeater $component, array $arguments, Set $set, Get $get) {
                $state = $component->getState();
                $item = $state[$arguments["item"]];

                $response = Http::baseUrl(config("api.url"))
                    ->timeout(120)
                    ->connectTimeout(120)
                    ->withHeaders([
                        "X-API-KEY" => config("api.token")
                    ])->attach(['code' => 15])->post("/tasks/image");
                $data = json_decode($response->body(), true);

                $offset = array_search($arguments["item"], array_keys($state));
                $state = array_slice($state, 0, $offset, true) + [
                        (string)Str::uuid() => [
                            "temporary" => $data['content'],
                            "beforeFilename" => $item["filename"],
                        ]
                    ] + array_slice($state, $offset, preserve_keys: true);
                unset($state[$arguments["item"]]);
                $set("photos", $state);

                Notification::make()
                    ->title("Successful")
                    ->success()
                    ->persistent()
                    ->send();
            });

**Reply by leandrocfe** [reply]
Is it updating the fields in the repeater as you need?

**Reply by devEcommercePL** [reply]
Yes, everything works correctly here. As I wrote earlier, by clicking several buttons at the same time, the actions are synchronous - the result of all of them will appear after the last one is performed, not after each action.
